src/backend/processor/diagnostics.ts

```diff
@@ -0,0 +1,196 @@
+import { Event, EventEmitter, ExtensionContext, TextEditor, Uri, window } from 'vscode';
+import { parseDiagnostics } from '../parser';
+import { CheckerMetadata, DiagnosticEntry, DiagnosticFile } from '../types';
+import { ExtensionApi } from '../api';
+
+/**
+ * API interface that provides Diagnostics data.  
+ * Access the active instance via ExtensionApi.
+ */
+export class DiagnosticsApi {
+    private _openedFiles: string[] = [];
+
+    /** 
+     * Content based on the contents of loaded .plist files.
+     * Key: Source code file, value: .plist analysis files
+     */
+    private _diagnosticSourceFiles: Map<string, string[]> = new Map();
+
+    /**
+     * Loaded diagnostic files.
+     * Key: .plist analysis file, value: file's contents
+     */
+    private _diagnosticEntries: Map<string, DiagnosticFile> = new Map();
+
+    private _selectedEntry?: {
+        file: string,
+        idx: number
+    };
+    public get selectedEntry(): {
+        position: {readonly file: string, readonly idx: number},
+        diagnostic: DiagnosticEntry
+    } | undefined {
+        if (this._selectedEntry) {
+            const activeFile = this._diagnosticEntries.get(this._selectedEntry.file);
+            const diagnostic = activeFile?.diagnostics[this._selectedEntry.idx];
+
+            if (diagnostic) {
+                return {
+                    position: this._selectedEntry!,
+                    diagnostic
+                };
+            }
+        }
+
+        return undefined;
+    }
+
+    public setActiveReport(position: {file: string, idx: number} | undefined) {
+        if (position) {
+            const activeFile = this._diagnosticEntries.get(position.file);
+            const diagnostic = activeFile?.diagnostics[position.idx];
+
+            if (diagnostic) {
+                this._selectedEntry = position;
+                return;
+            }
+        }
+
+        this._selectedEntry = undefined;
```

Comment:
Why don't we assign `null` to this variable instead of `undefined`. `undefined` means that the variable is declared but no value is assigned to it.
```suggestion
        this._selectedEntry = null;
```
Also we can initialize this variable to null to identify that no entry is selected.

Reply:
In Typescript, using `undefined` is just more convenient in variables (eg. with the ``variable?: Type`` syntax for declaration.) With null, ``Type | null`` has to be explicitly written out in every type definition.

I did use it as a return value though, as there's no syntax for an optional return.